nixos/modules/services/network-filesystems/openafs-server.nix

```diff
@@ -0,0 +1,146 @@
+    end
+    bnode dafs dafs 1
+    parm ${openafsPkgs}/libexec/openafs/dafileserver -vattachpar 128 -vhashsize 11 -cb 1000000
+    parm ${openafsPkgs}/libexec/openafs/davolserver
+    parm ${openafsPkgs}/libexec/openafs/salvageserver
+    parm ${openafsPkgs}/libexec/openafs/dasalvager
+    end
+  '';
+in
+{
+  options = {
+
+    services.openafsServer = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to enable the OpenAFS server.";
+      };
+
+      adminUsers = mkOption {
+        type = types.listOf types.str;
+        default = [];
+        description = "List of OpenAFS users that can administer this server.";
+      };
+
+      advertisedAddresses = mkOption {
+        default = [];
+        type = types.listOf types.str;
+        description = "List of IP addresses this server is advertised under. See NetInfo(5)";
+      };
+      bosConfig = mkOption {
+        type = types.package;
+        default = bosConfig;
+        defaultText = ''
+          restarttime 16 0 0 0 0
+          checkbintime 3 0 5 0 0
+          bnode simple vlserver 1
+          parm ''${cfg.openafsPkgs}/libexec/openafs/vlserver
+          end
+          bnode simple ptserver 1
+          parm ''${cfg.openafsPkgs}/libexec/openafs/ptserver -db /var/openafs/prdb -restricted
+          end
+          bnode dafs dafs 1
+          parm ''${cfg.openafsPkgs}/libexec/openafs/dafileserver -vattachpar 128 -vhashsize 11 -cb 1000000
+          parm ''${cfg.openafsPkgs}/libexec/openafs/davolserver
+          parm ''${cfg.openafsPkgs}/libexec/openafs/salvageserver
+          parm ''${cfg.openafsPkgs}/libexec/openafs/dasalvager
+          end
+        '';
+        description = "The server configuration file specifying the started servers and options";
+      };
+
+      cellName = mkOption {
+        type = types.str;
+        default = "grand.central.org";
+        description = "Server's cell name.";
+      };
+
+      cellServers = mkOption {
+        type = types.listOf (mkOptionType {
+          name = "IPNameMap";
+          description = "map of IP address to DNS name";
+          check = x: isAttrs x && x ? address && x ? hostname;
+          merge = mergeOneOption;
+        });
+
+        default = [];
+        description = "IP addresses of database servers that serve this cell.";
+        example = literalExample ''
+          [ { address = "10.0.0.3"; hostname = "afsdb.example.org"; } ]
+        '';
+      };
+
+      openafsPkgs = mkOption {
+        type = types.package;
+        default = config.boot.kernelPackages.openafs;
+        description = "The package to use for the openafs servers";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ openafsPkgs ];
+
+    system.activationScripts.openafsServer = stringAfter
+    [ (if config.services.openafsClient.enable then "openafs" else "etc") ] ''
```

Comment:
It could be done in `preStart`. I wasn't able yet to figure out, what to do in `activationScripts` and when to use `preStart` of a service. Is `preStart` preferred for services?